main.tf

```diff
@@ -41,59 +29,10 @@ data "aws_caller_identity" "current" {
 data "aws_region" "current" {
 }
 
-resource "aws_kms_key" "s3_bucket_kms_key" {
-  count = var.kms_alias != "" && length(var.whitelist_ip) == 0 && length(var.whitelist_vpc) == 0 && var.website_hosting == "false" ? 1 : 0
-
-  description = "A kms key for encrypting/decrypting S3 bucket ${var.name}"
-  policy      = var.kms_key_policy != "" ? var.kms_key_policy : data.aws_iam_policy_document.kms_key_policy_document.json
-
-  tags = merge(
-    var.tags,
-    {
-      "Name" = format("%s-%s", var.environment, var.name)
-    },
-    {
-      "Env" = var.environment
-    },
-  )
-}
-
-resource "aws_kms_alias" "s3_bucket_kms_alias" {
-  count = var.kms_alias != "" && length(var.whitelist_ip) == 0 && length(var.whitelist_vpc) == 0 && var.website_hosting == "false" ? 1 : 0
-
-  name          = "alias/${var.kms_alias}"
-  target_key_id = aws_kms_key.s3_bucket_kms_key[0].key_id
-}
-
-resource "aws_kms_key" "s3_bucket_kms_key_whitelist" {
-  count = var.kms_alias != "" && length(var.whitelist_ip) != 0 && length(var.whitelist_vpc) == 0 && var.website_hosting == "false" ? 1 : 0
-
-  description = "A kms key for encrypting/decrypting S3 bucket ${var.name}"
-  policy      = var.kms_key_policy != "" ? var.kms_key_policy : data.aws_iam_policy_document.kms_key_policy_document_whitelist.json
-
-  tags = merge(
-    var.tags,
-    {
-      "Name" = format("%s-%s", var.environment, var.name)
-    },
-    {
-      "Env" = var.environment
-    },
-  )
-}
-
-resource "aws_kms_alias" "s3_bucket_kms_alias_whitelist" {
-  count = var.kms_alias != "" && length(var.whitelist_ip) != 0 && length(var.whitelist_vpc) == 0 && var.website_hosting == "false" ? 1 : 0
-
-  name          = "alias/${var.kms_alias}"
-  target_key_id = aws_kms_key.s3_bucket_kms_key_whitelist[0].key_id
-}
-
-resource "aws_kms_key" "s3_bucket_kms_key_whitelist_vpc" {
-  count = var.kms_alias != "" && length(var.whitelist_ip) == 0 && length(var.whitelist_vpc) != 0 && var.website_hosting == "false" ? 1 : 0
-
-  description = "A kms key for encrypting/decrypting S3 bucket ${var.name}"
-  policy      = var.kms_key_policy != "" ? var.kms_key_policy : data.aws_iam_policy_document.kms_key_with_whitelist_vpc_policy_document.json
+resource "aws_kms_key" "this" {
+  description         = "A kms key for encrypting/decrypting S3 bucket ${var.name}"
+  enable_key_rotation = var.cmk_enable_key_rotation
+  policy              = var.kms_key_policy != "" ? var.kms_key_policy : data.aws_iam_policy_document.kms_key_policy_document.json
```

Comment:
Encryption is no longer optional, so the bucket will be encrypted with a KMS key or the default AWS S3 key.
KMS encryption is optional, but so far I haven't found a terraform that makes this optional and usable where it needs to be.
I think that having a single bucket resource (to avoid failed plans and terraform state refactoring) is more important than creating superfluous resources, so this resource not being optional is a consequence of that. I believe that it's the only one

Reply:
Refactored further to avoid having unnecessary resources being created